text/0000-vhdl-emitter.md

```diff
@@ -0,0 +1,138 @@
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+In the FIRRTL compiler, there are multiple different `Emitter`s that you can choose from. These are implemented as subclasses of `firrtl.Transform`, i.e., mathematical transformations on a `firrtl.CircuitState` (`(a: CircuitState) => CircuitState`). Concretely, each emitter transform reads in a FIRRTL circuit, does some pre-processing on it, and generates a `firrtl.EmittedAnnotation` which contains the emitted circuit as a `String`. The underlying types used in this process are shown below:
+
+```scala
+sealed abstract class EmittedComponent {
+  def name: String
+  def value: String
+  def outputSuffix: String
+}
+
+sealed trait EmittedAnnotation[T <: EmittedComponent] extends NoTargetAnnotation {
+  val value: T
+}
+```
+
+An emitter transform does not write-back the emitted circuit to disk (there is no IO side-effect). Instead, the final `firrtl.options.Phase` of the FIRRTL compiler is a `firrtl.stage.phases.WriteEmitted` phase that looks for any `EmittedAnnotation`s and dumps them to the location specified by a `firrtl.options.TargetDirAnnotation`.
+
+FIRRTL provides the following emitters:
+
+- `ChiselEmitter`
+- `HighFirrtlEmitter`
+- `MiddleFirrtlEmitter`
+- `LowFirrtlEmitter`
+- `MinimumVerilogEmitter`
+- `VerilogEmitter`
+- `SystemVerilogEmitter`
+- `MinimumVHDLEmitter`
+- `VHDLEmitter`
+
+All of these emitters are implemented using the approach above. They are `Transform`s that generated one or more `EmittedAnnotation`s which will be written back when the `WriteEmitted` phase runs. Some emitters are pass-through and simply take the existing in-memory circuit representation and convert it to a `String`. These include the `{Chisel, High, Middle, Low}` emitters. The Verilog and VHDL emitter families do Verilog or VHDL-specific pre-processing and generate `EmittedAnnotation`s holding strings of Verilog or VHDL. "Minimal" variants do the bare-minimum of required pre-processing, but do not do additional, optional optimizations (e.g., constant propagation, primitive operator grouping, etc.).
+
+A user can specify which emitter (or emitters) they want using an explicit `RunFirrtlTransformAnnotation` requesting that a specific emitter run or they can use the FIRRTL fat jar as following:
+
+```bash
+# To emit a single file containing the whole design in VHDL
+firrtl -E vhdl
+
+# To emit one-file-per-module of Verilog
+firrtl -e verilog
+
+# To emit both Verilog and VHDL
+firrtl -E verilog -E vhdl
```

Comment:
I'm in agreement with this. The starting point for a Verilog, VHDL, or some other language emitter would be Low FIRRTL. Currently language-specific optimizations are not saved and are only used to help generate the emitted `String` in the target language. The actual emitter transform is only adding annotations that include the emitted `String`.

The existing path from Low FIRRTL to "Low FIRRTL Optimized and ready for Verilog conversion" is here to show some flavor of what's done: https://github.com/freechipsproject/firrtl/blob/master/src/main/scala/firrtl/Emitter.scala#L949. None of this is saved to the IR, though. I'd expect a similar set of passes would be necessary for VHDL.

Scoping to one specific Version of VHDL would make sense. There's no necessity on why FIRRTL couldn't support additional versions eventually. Related to this, the `SystemVerilogEmitter` is really just the `VerilogEmitter` with a different file suffix. Eventually, this could include other custom transforms (e.g., interface construction) that specialize better to the features of that language.

Fortunately, [removal of collision with a set of keywords can be made parametric](https://github.com/freechipsproject/firrtl/blob/master/src/main/scala/firrtl/transforms/RemoveKeywordCollisions.scala#L15). I expect that there's similar optimization transforms that can be written to be reusable across language backends.

Good thoughts, thanks!